README.md
# Gare la plus proche

Find the nearest train station from the train network in France (SNCF).

Available on: http://www.garelaplusproche.fr

Thanks to [laurencebeillaux](https://github.com/laurencebeillaux) who created all the website UI/UX.

## Run 

```
npm install 
npm start
```
Note : tested with Node 16.x. 

## Tests 

```
npm test
```

## Contributing 

Feel free to create an [issue](https://github.com/glae/garelaplusproche.fr/issues) for bug or enhancement.

## Data sources

There are two inputs: 

- train stations list: https://ressources.data.sncf.com/explore/dataset/referentiel-gares-voyageurs/ (minimized in [source code](stations-list.js#L1) with some [`jq` commands](tools/some_useful_commands_to_minimize_stations_file.sh))
- addresses/cities search API: https://adresse.data.gouv.fr/api

Two outputs: 

- map: http://leafletjs.com/ (and [OpenStreetMap](http://openstreetmap.org))
- to buy tickets: http://trainline.fr ("URL-friendly" alternative to https://www.sncf-connect.com/) 

## Motivations

I was in vacation in the french coutryside on 10/09/2017. I had to drive a friend to the nearest train station, I fought on the web several minutes to get the right information. 

I wish it could have been easier. That's why I created this simple website.

If you can read french you can get a longer explanation [here](doc/README-motivations-fr.md).


CHANGELOG.md
# Changelog

## v1.2.0 - 2023-10-06
- #30 improve search by city
- #5 text for mobile/desktop

## v1.1.1 - 2023-09-16
- minor changes, fix mobile layout

## v1.1.0 - 2023-09-14 
- Add #24 (permalink sharing)

## v1.0.0 - 2017-11-08
First public release 

main.test.js
const {stations, findNearestStations} = require('./main.js');
import test from 'ava';


test('we should have 3002 stations', t => {

  t.is(stations.length, 3002);

});

test('Anetz should have nearest stations', t => {

  var sta = findNearestStations(stations, 47.383333, -1.100000);
	t.is(Object.keys(sta[0])[0], "Mauves-sur-Loire");
  t.is(Object.keys(sta[1])[0], "Oudon");

});

test('Saint-Aubin-de-Luigné should have nearest stations', t => {

  var sta = findNearestStations(stations, 47.340512, -0.675747);
  t.is(sta.length, 10);

});


run_tests.sh
#!/bin/bash
#use https://github.com/avajs/ava
npm test -- --verbose tests/main.test.js


tests/main.test.js
const {stations, findNearestStations, proposalsFrom} = require('../main.js');
const fs = require('fs');
import test from 'ava';

test('we should have 3002 stations', t => {
  t.is(stations.length, 3002);
});

test('Anetz should have nearest stations', t => {
  const foundStations = findNearestStations(stations, 47.383333, -1.100000);
	t.is(Object.keys(foundStations[0])[0], "Mauves-sur-Loire");
  t.is(Object.keys(foundStations[1])[0], "Oudon");
});

test('Saint-Aubin-de-Luigné should have nearest stations', t => {
  const foundStations = findNearestStations(stations, 47.340512, -0.675747);
  t.is(foundStations.length, 15);
});

test('As I user, I should get no autocomplete address when I type "Unkown"', t => {
  const proposals = proposalsFrom(userInput('Unknown'));
  t.is(proposals.length, 0);
});


test('As I user, I should get autocomplete addresses when I type "Le Man"', t => {
  t.deepEqual(labelsFromInput(  'Le Man'),
                                ['Le Mans (72)', 'Le Manoir (27)', 'Le Manoir (14)', 'Boulazac Isle Manoire (24)']);
});


test('______________________________________________ when I type "Par"', t => {
  t.deepEqual(labelsFromInput(  'Par'),
                                [ 'Paris (75)',
                                  'Parthenay (79)',
                                  'Parignargues (30)',
                                  'Parpeville (02)',
                                  'Parey-Saint-Césaire (54)' ]);
});

test('______________________________________________ when I type "Pe"', t => {
  t.deepEqual(labelsFromInput(  'Pe'),
                              [ 'Saint-Pé-de-Bigorre (65)',
                                'Saint-Pée-sur-Nivelle (64)',
                                'Saint-Pé-de-Léren (64)',
                                'Saint-Pé-d\'Ardet (31)',
                                'Notre-Dame-du-Pé (72)' ]);
});


test('______________________________________________ when I type "R"', t => {
  t.deepEqual(labelsFromInput(  'R'),
                              [ 'Rue (80)',
                            'Vatteville-la-Rue (76)',
                            'La Rue-Saint-Pierre (76)',
                            'Les Rues-des-Vignes (59)',
                            'Saulon-la-Rue (21)' ]);
});

test('______________________________________________ when I type "Saint-Au"', t => {

  t.deepEqual(labelsFromInput(  'Saint-Au'),
                                [ 'Saint-Augustin (62)',
                              'Saint-Aubin (02)',
                              'Saint-Aubert (59)',
                              'Saint-Aubin (62)' ]);
});


test('______________________________________________ when I type "Tours"', t => {
  t.deepEqual(labelsFromInput(  'Tours'),
                                [ 'Tours (37)', 'Tours-en-Savoie (73)', 'Tours-sur-Marne (51)' ]);
});


test('______________________________________________ when I type "Vair"', t => {
  t.deepEqual(labelsFromInput(  'Vair'),
                              [ 'Vaires-sur-Marne (77)',
                            'Vairé (85)',
                            'Vaire (25)',
                            'Vair-sur-Loire (44)',
                            'Vaire-sous-Corbie (80)' ]);
});


test('______________________________________________ when I type "Mons"', t => {
  t.deepEqual(labelsFromInput(  'Mons'),
                              [ 'Mons (83)', 'Mons (30)', 'Mons (16)', 'Mons (31)', 'Mons (34)' ]);
});


function labelsFromInput(word){
  return proposalsFrom(userInput(word)).map(n=>n.label);
}

function userInput(word){
  return JSON.parse(fs.readFileSync('tests/adresse.data.gouv.fr_api_samples/' + word + '.json').toString());
}


stations-functions.js
/* 
Finds the nearest train stations from GPS coords. Recursively called to find 15 stations max. 

This is a compromise between:
 - a JS limitation between 'findNearestStations' recursive iterations (possible error: Maximum call stack size exceeded):
   0.001 radius seems to be fine (most precise coords : [48.81550666129636,1.876934489891097])
 - precision/circle/result exhaustiveness (min. 15 results requirement seem to compensate the radius precision issue)

 Variable 'stations' is defined in 'stations.js' file (TODO refactor as modules/etc)
*/
function findNearestStations(lat, long, searchRadius = 0.001) {
  const minLat = lat - searchRadius;
  const maxLat = lat + searchRadius;
  const minLong = long - searchRadius;
  const maxLong = long + searchRadius;

  const nearestStations = stations.filter(function (s) {
    const lat = latitude(s);
    const long = longitude(s);
    if (lat >= minLat && lat <= maxLat && long >= minLong && long <= maxLong) {
      return true;
    } else {
      return false;
    }
  });

  if (nearestStations.length < 15) {
    return findNearestStations(lat, long, searchRadius + 0.001);
  } else {
    return nearestStations;
  }
}

function stationName(e) {
  return Object.keys(e)[0];
}

function latitude(e) {
  return Object.keys(e).map(k => e[k])[0][0];
}

function longitude(e) {
  return Object.keys(e).map(k => e[k])[0][1];
}

function createProposals(input, onAPIResultComplete) {
  const cleanInput = input.trim();
  var request = new XMLHttpRequest();
  request.onreadystatechange = function (response) {
    if (request.readyState === 4) {
      if (request.status === 200) {
        onAPIResultComplete(proposalsFrom(JSON.parse(request.responseText)));
      }
    }
  };
  if (cleanInput.length != 0) {
    request.open('GET', 'https://api-adresse.data.gouv.fr/search/?q=' + cleanInput + '&type=municipality', true);
    request.send();
  }
}

function proposalsFrom(jsonAddresses) {
  let uniqueAddresses = [];
  if (jsonAddresses.features == null) return uniqueAddresses;
  const addresses = jsonAddresses.features.map(function (address) {
    return {
      label: (address.properties.name + ' (' + address.properties.postcode.substring(0, 2) + ')'),
      lat: address.geometry.coordinates[1],
      lon: address.geometry.coordinates[0]
    }
  });
  addresses.forEach(function (a) {
    if (!uniqueAddresses.filter(u => u.label == a.label).length > 0) {
      uniqueAddresses.push(a);
    }
  });
  return uniqueAddresses;
}

if (typeof module !== 'undefined') { //for unit tests
  module.exports.stations = stations;
  module.exports.findNearestStations = findNearestStations;
  module.exports.proposalsFrom = proposalsFrom;
}


index.js
var map;
var lastProposedCities = [];
checkUrlParams();
initSearchByCityInput();

function copyLink() {
    navigator.clipboard.writeText(window.location.href);
    document.getElementById('copied').innerText = "Lien copié dans le presse-papier !";
}

function initSearchByCityInput() {
    const input = document.getElementById("station-search-container-by-city-input");
    input.addEventListener("input", function (e) {
        function onAPIResultComplete(proposals) {
            const cities = document.getElementById('filtered-cities');
            cities.innerHTML = "";
            lastProposedCities = proposals;
            proposals.forEach(
                function (city) {
                    var option = document.createElement('option');
                    option.value = city.label;
                    cities.appendChild(option);
                });
        }
        document.getElementById("error-city").innerHTML = '';
        createProposals(input.value, onAPIResultComplete);
    });
}

function checkUrlParams() {
    var params = new URLSearchParams(window.location.search);
    if (params.has('lat') && params.has('long')) {
        const lat = parseFloat(params.get('lat'));
        const long = parseFloat(params.get('long'));
        searchByCoords(lat, long);
    }
}

function hideSearchFields() {
    document.getElementById("search-btn-container").className += 'active-button';
    document.getElementById("train-station").className += 'loader';
    document.getElementById("stations").className += "station-element";
    document.getElementById("station-search-container-by-city").className += 'active-button';
}

function searchByGeolocation() {
    hideSearchFields();
    var warning = document.createElement("p");
    warning.innerText = 'Si rien ne se passe, c\'est peut-être que vous n\'avez pas accepté la géolocalisation.';
    setTimeout(function () {
        document.getElementById("trainLoader-buttonSearch-container").appendChild(warning);
    }, 6000);
    var geoSuccess = function (p) {
        const userLat = p.coords.latitude;
        const userLong = p.coords.longitude;
        search(userLat, userLong);
    };
    navigator.geolocation.getCurrentPosition(geoSuccess);
}

function searchByCity() {
    console.log("searchByCity")
    const cityInput = document.getElementById("station-search-container-by-city-input").value;
    console.log(cityInput);
    const selectedCity = lastProposedCities.filter(u => u.label == cityInput.trim())[0];
    console.log(selectedCity);
    if (selectedCity) {
        hideSearchFields();
        search(selectedCity.lat, selectedCity.lon, selectedCity.label);
    }
}

function searchByCoords(userLat, userLong) {
    hideSearchFields();
    search(userLat, userLong);
}

function search(userLat, userLong, positionLabel = 'Votre position') {
    if (window.location.search === '') {
        history.replaceState(null, '', `/?lat=${userLat}&long=${userLong}`);
    }
    var mapContainerElement = document.getElementById('mapcontainer');
    mapContainerElement.innerHTML = "<div id='mapid'></div>";
    mapContainerElement.className += "map-container-element";
    map = L.map('mapid').setView([userLat, userLong], 11);
    L.tileLayer('http://{s}.tile.osm.org/{z}/{x}/{y}.png').addTo(map);
    const foundStations = findNearestStations(userLat, userLong);
    var trainIcon = new L.Icon({
        iconUrl: './img/tchou.png',
        iconSize: [40, 40]
    });
    var homeIcon = new L.Icon({
        iconUrl: './img/yourPosition.png',
        iconSize: [40, 51]
    });
    foundStations.forEach(function (e) {
        L.marker([latitude(e), longitude(e)], {
            icon: trainIcon
        }).addTo(map).bindPopup("<p style='font-size:150%'>" +
            "<a target='_blank' style='color: #D35400; text-decoration: none;' href='https://www.openstreetmap.org/?mlat=" + latitude(e) + "&mlon=" + longitude(e) + "'>" + stationName(e) + "</a>" +
            "<br><a href='https://www.trainline.fr/search/" + encodeURI(stationName(e).replace(/'/g, "%27")) + "/' target='_blank' style='color: #D35400; font-weight: 700; text-decoration: none;'>Partir avec trainline</a>" +
            "<br>" +
            "</p>"
        );
    });
    L.marker([userLat, userLong], {
        icon: homeIcon
    }).addTo(map).bindPopup('<p style="font-size:150%">' + positionLabel + '</p>');
    var showedStations;
    var others = ""
    if (foundStations.length >= 24) {
        showedStations = foundStations.slice(0, 23);
        others = "<li><b>et d'autres encore...<b></li>"
    } else {
        showedStations = foundStations;
    }
    stationsHtml = document.getElementById('stations');
    stationsHtml.innerHTML = '' +
        '<h2>Voici les gares trouvées :</h2><ul>' +
        showedStations.map(gare => '<li>' + stationName(gare) + '</li>').join('') +
        others + '</ul>';
    document.getElementById("share-link").style.visibility = 'visible';
    var mobileHeaderTitle = document.getElementById("header");
    mobileHeaderTitle.className += "mobile";
}
